### ISSUE TITLE
zip-download of a gallery fails after approx. 65Mb

### ISSUE THREAD

**Issue by SvenMertin** [LycheeOrg/Lychee#1434]
Hi there, 
I cannot figure out, why in my case the zip-download is not possible. I tested three different galleries and in each case the download fails after reaching 65Mb. No file is stored.

### Steps to reproduce the issue
1. Go to gallery
2. Click on Download
3. Wait Download size to reach approx. 65Mb

Any ideas? Thank you :-)

### Diagnose output
    Diagnostics
    -------
    Warning: Dropbox import not working. dropbox_key is empty.

    System Information
    --------------
    Lychee Version (git):            master (f39be7a) -- Up to date (5 minutes ago).
    DB Version:                      4.5.2
    
    composer install:                --no-dev
    APP_ENV:                         production
    APP_DEBUG:                       false
    
    System:                          Linux
    PHP Version:                     8.1.8
    PHP User agent:                  Lychee/4 (https://lycheeorg.github.io/)
    Timezone:                        Europe/Berlin
    Max uploaded file size:          128M
    Max post size:                   128M
    Max execution time:              200
    MySQL Version:                   10.3.35-MariaDB-1:10.3.35+maria~bionic-log
    
    exec() Available:                yes
    Imagick Available:               1
    Imagick Enabled:                 1
    Imagick Version:                 1687
    GD Version:                      2.3.3

    Config Information
    --------------
    version:                         040502
    check_for_updates:               0
    sorting_photos_col:              taken_at
    sorting_photos_order:            ASC
    sorting_albums_col:              max_taken_at
    sorting_albums_order:            ASC
    imagick:                         1
    skip_duplicates:                 0
    small_max_width:                 0
    small_max_height:                360
    medium_max_width:                1920
    medium_max_height:               1080
    lang:                            de
    layout:                          1
    image_overlay_type:              desc
    default_license:                 none
    compression_quality:             90
    full_photo:                      1
    delete_imported:                 0
    Mod_Frame:                       1
    Mod_Frame_refresh:               30
    thumb_2x:                        1
    small_2x:                        1
    medium_2x:                       1
    landing_page_enable:             0
    landing_owner:                   Caro
    landing_title:                   
    landing_subtitle:                
    landing_facebook:                
    landing_flickr:                  
    landing_twitter:                 
    landing_instagram:               
    landing_youtube:                 
    landing_background:              
    site_title:                      Online-Gallerie
    site_copyright_enable:           1
    site_copyright_begin:            2022
    site_copyright_end:              2022
    additional_footer_text:          Online-Gallerie
    display_social_in_gallery:       0
    public_search:                   0
    SL_enable:                       0
    SL_for_admin:                    0
    public_recent:                   0
    recent_age:                      1
    public_starred:                  0
    downloadable:                    0
    photos_wraparound:               1
    map_display:                     0
    zip64:                           0
    map_display_public:              0
    map_provider:                    Wikimedia
    force_32bit_ids:                 0
    map_include_subalbums:           0
    update_check_every_days:         3
    has_exiftool:                    0
    share_button_visible:            0
    import_via_symlink:              0
    has_ffmpeg:                      1
    location_decoding:               0
    location_decoding_timeout:       30
    location_show:                   1
    location_show_public:            0
    rss_enable:                      0
    rss_recent_days:                 7
    rss_max_items:                   100
    prefer_available_xmp_metadata:   0
    editor_enabled:                  1
    lossless_optimization:           0
    swipe_tolerance_x:               150
    swipe_tolerance_y:               250
    local_takestamp_video_formats:   .avi|.mov
    log_max_num_line:                1000
    unlock_password_photos_with_url_param: 0
    nsfw_visible:                    1
    nsfw_blur:                       0
    nsfw_warning:                    0
    nsfw_warning_admin:              0
    map_display_direction:           1
    album_subtitle_type:             oldstyle
    upload_processing_limit:         4
    public_photos_hidden:            1
    new_photos_notification:         0
    legacy_id_redirection:           1

**Comment by nagmat84** [comment]
Are there any related warnings or errors in the log which might guide us?

At the moment I cannot think of any plausible bug inside Lychee why Lychee should fail after 65MB. But 65MB smells like 64MB and maybe there is some kind of upper limit configured somewhere which is hit? (Although I have never heard of an upper download limit.)

**Comment by SvenMertin** [comment]
Unfortunately, the error logs of nginx and lychee are not showing anything as they are empty. This is the same for the lynchee logs within the web interface. The access log of lychee, however, is showing the following in the exact second, the download fails (edited ip):

`2.xx.xx.59 - - [03/Aug/2022:08:04:09 +0200] "GET /gallerie/api/Album::getArchive?albumIDs=i_Eao2ovGnXTv3KL9iqtBP1l HTTP/2.0" 200 58172959 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Safari/537.36"
`
But this is not really helping either in my opinion.

I was not 100% sure if the download cut is really related to 64Mb or if its just time based and because of the download rate at the approx. same time. I just started a download with reduced download rate and it seems to be time-related rather than size-related. The download fails at roughly 24s.

**Comment by nagmat84** [comment]
Your diagnostic output says `Max execution time: 200`. Lychee tries to reset the execution time counter after every file which has been ZIPped, but the permission to do so might be disabled by configuration. (I am not sure if we log warning in this case, i.e. of Lychee tries to reset the execution timer but fails, or if we proceed silently.)

Could you try to set `max execution time` to a different value and try again?

**Comment by SvenMertin** [comment]
Hi nagmat84, 
thank you for your active help :-) I really appreciate it!

So I set the `max_execution_time` to 3600, restarted php and tested it again: Unfortunately there is no change :-/

Another info: I have a nextcloud instance in parallel which runs on the same webserver. With this there are no problems with downloading even very large zip folders (but it uses an older php release, i guess 7.4).

**Comment by nagmat84** [comment]
No problem, however, I am off to work now.

**Comment by SvenMertin** [comment]
What is remarkable:
- Downloading a gallery as zip using lychee will lead to 90%-99% cpu usage with php8.2 in my case. 
- Downloading a folder as zip from my nextcloud will lead to 5%-10% cpu usage with php7.4.

both running on the same nginx instance.

I am not a regular linux user. Is there some kind of process monitor which kills automatically a process if the cpu usage is very high over a specific time? (I am using Ubuntu 18.04.6 LTS)

**Comment by kamil4** [comment]
Can you check if setting `zip64` in the advanced settings changes anything? It probably won't (it shouldn't make a difference until a 4 GB archive or 64K files), but it's the easiest thing to try first...

What's your server like? It it shared hosting managed by somebody else, or a VPS, or..?

PHP will normally kill web processes using more than `max_execution_time` of CPU time but even with the Lychee default of 200s, you should be fine here. There could also be a limit elsewhere in your environment I suppose. Is your web server behind a proxy?

Lychee tries to compress images it stores in the zip file, which could explain the high CPU usage (nextcloud could be storing them uncompressed; you can check the compression method in the generate zip file). We were planning to disable the compression but this was held up by a pending refactoring that's been merged since then; @nagmat84, can we disable the compression now? It would be interesting to see if that would make a difference here...

Is your gallery public? If so, could you share the URL of at least one problematic album with us? We won't be able to see what might be going on on the server side, but at least we should be able to eliminate the client side as a possible problem...

**Comment by nagmat84** [comment]
> can we disable the compression now? It would be interesting to see if that would make a difference here...

The same thought hit me during the day. We talked about that in the context of issue https://github.com/LycheeOrg/Lychee/issues/1295 issue, too. I promised that I would add the feature to disable compression when ZIP files are created from streams. But that got lost. Sorry!

I will create a PR this evening.

**Comment by nagmat84** [comment]
Here: https://github.com/LycheeOrg/Lychee/pull/1436

**Comment by SvenMertin** [comment]
Hi, thank you for your suggestions :-)

- disabling zip64 didn't change anything
- I switched to the [disable_zip_compression_by_option](https://github.com/LycheeOrg/Lychee/tree/disable_zip_compression_by_option) branch and tested if anything changes by disabling the compression (-1) in the settings but unfortunately, this also didn't change anything in my case :-( What is interesting in addition is that the cpu load also didn't change! So I didn't check all changes in this branch, maybe the function is just not yet implemented but only the settings visibility?
- My gallery is not public but I can set up a public gallery and will provide a link (I think tomorrow)

Thank you very much for your nice support!
Regards
Sven

**Comment by kamil4** [comment]
@SvenMertin Give #1436 another try; @nagmat84 pushed some further fixes there so that now it actually works as intended :smile:.

**Comment by SvenMertin** [comment]
@kamil4 I tried #1436 with these results:
The download again fails after ~25s if compression is set to a value >=0 and I could also not notice a real difference in my cpu load. If the compression is set to -1 the download fails immedately. So this didn't solve my problems. I am preparing a public gallery and will provide you the link ;-)

**Comment by SvenMertin** [comment]
So here you have a sample gallery (I used a url shortener, so my domain cannot being found by search engines):

[https://bit.ly/3PZZZdj](https://bit.ly/3PZZZdj)

**In my case the download is working just without problems?!** 

The difference between the sample gallery and my 'real' galleries is, that the real galleries contain images with a size up to 25MB while the sample images are much smaller (max 7MB).

Could this be a hint? Does the file size have impact on my problem?

Thanks :-)

**Comment by qwerty287** [comment]
This works for me. However, it looks lime there's no `content-length` header which makes it a bit hard... That's something else we should fix.

**Comment by nagmat84** [comment]
> This works for me. However, it looks lime there's no `content-length` header which makes it a bit hard... That's something else we should fix.

We can't. The response is a streamed response, i.e. the beginning of the archive is already being sent to the client, while more files are added to the archive. We don't know the total size in advance.

The alternative would be to compress and create the whole archive first and then send it to the client. But we had this in the past and it may lead to client-side timeouts when the client does not receive an answer while the server is still creating the archive.

**Comment by qwerty287** [comment]
OK, sorry didn't know this.

**Comment by kamil4** [comment]
@SvenMertin Yeah, the gallery you posted works for me as well. Can you set up one that does not work for you?

**Comment by nagmat84** [comment]
Works for me, too. I have an average download rate of 7 MBit/s which I deem sufficiently good. That's nothing I would complain about.

**Comment by SvenMertin** [comment]
Hey!
So after many things I have tried I went with a new and clean reinstall of lychee. I noticed then, that I can no longer provoke a download fail especially using large sample images. Therefore I reuploaded my private gallery which I wasn't able to download before. 

Now it's working with default settings and even the cpu load is only around 50-70%. 

I don't know what's the difference as I did the installation as before, but obviously something must have gone wrong before?

Sorry for taking your time for this :-/ But I really appreciate the support.
